// powercord/components/ContextMenu/Button.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2019  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { React, getModule } = require('powercord/webpack');
const { getOwnerInstance, waitFor } = require('powercord/util');

const ItemGroup = require('./ItemGroup.jsx');

let classes = null;
let appOwnerInstance = null;
setImmediate(async () => {
  classes = {
    ...await getModule([ 'scrollbar', 'scrollerWrap' ]),
    ...await getModule([ 'itemToggle', 'checkbox' ])
  };

  waitFor(`.${(await getModule([ 'app' ])).app.replace(/ /g, '.')}`)
    .then(app => appOwnerInstance = getOwnerInstance(app));
});

module.exports = class ButtonItem extends React.PureComponent {
  onClick () {
    if (this.props.disabled) {
      appOwnerInstance.shake(600, 5);
    } else {
      this.props.onClick();
    }
  }

  render () {
    const button = (
      <div
        className={[
          classes.item,
          this.props.image ? classes.itemImage : null,
          classes.clickable,
          this.props.disabled ? classes.disabled : null
        ].filter(Boolean).join(' ')}
        role='button'
        onClick={this.onClick.bind(this)}
      >
        <span className={classes.label} style={{ color: this.props.highlight }}>
          {this.props.name}
        </span>

        {this.props.image
          ? this.getButtonImage()
          : <div className={classes.hint}>
            {this.props.hint}
          </div>}
      </div>
    );

    if (this.props.seperate) {
      return React.createElement(ItemGroup, {
        children: [ button ]
      });
    }

    return button;
  }

  getButtonImage () {
    return (
      this.props.image.startsWith('fa-')
        ? <div style={{ cursor: 'pointer !important' }}
          className={`${this.getFontAwesomePrefix(this.props.image)}
            ${this.props.image.replace(/-regular|-light|-brand/gi, '')} fa-fw`} />
        : <img alt className={[ classes.image, this.props.className || null ].filter(Boolean).join(' ')}
          src={this.props.image} />
    );
  }

  getFontAwesomePrefix (str) {
    const suffixes = [ 'regular', 'light', 'brand' ];
    const stringParts = str.split('-');
    const suffix = suffixes.find(suffix => suffix === stringParts[stringParts.length - 1]);

    if (suffixes[suffix]) {
      return `fa${suffix.charAt(0).toLowerCase()}`;
    }

    return 'fas';
  }
};
