// powercord/components/ContextMenu/Checkbox.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2019  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { React, getModule } = require('powercord/webpack');

let oldState = null;

module.exports = class CheckboxItem extends React.PureComponent {
  constructor (props) {
    super(props);

    this.state = oldState || {
      itemClasses: props.itemClasses || '',
      checkboxClasses: '',
      checked: props.defaultState
    };
  }

  async componentDidMount () {
    const itemClasses = (await getModule([ 'itemToggle', 'checkbox' ]));

    if (!this.state.itemClasses && !this.state.checkboxClasses) {
      this.setState({ itemClasses,
        checkboxClasses: (await getModule([ 'checkboxElement' ]))
      });
    }
  }

  componentWillUnmount () {
    oldState = this.state;
  }

  onClick () {
    const checked = !this.state.checked;

    this.props.onToggle(checked);
    this.setState({ checked });
  }

  render () {
    const { itemClasses, checkboxClasses } = this.state;
    const checkbox = (
      <div
        className={[ itemClasses.item, itemClasses.itemToggle ].join(' ')}
        role='button'
        onClick={() => this.onClick()}
      >
        <div className={itemClasses.label}>
          {this.props.name}
        </div>

        <div className={[ checkboxClasses.checkbox, itemClasses.checkbox ].join(' ')} role='button'>
          <div className={checkboxClasses.checkboxInner}>
            <input className={checkboxClasses.checkboxElement} type='checkbox' checked={this.state.checked}/>
            <span/>
          </div>
        </div>
      </div>
    );

    if (this.props.seperate) {
      return (
        <div className={`${itemClasses.itemGroup} seperated`}>
          {checkbox}
        </div>
      );
    }

    return checkbox;
  }
};
