// powercord/components/ContextMenu/Checkbox.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2019  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { React, getModule } = require('powercord/webpack');

const ItemGroup = require('./ItemGroup');

let classes = null;
setImmediate(async () => {
  classes = {
    itemCheckbox: (await getModule([ 'itemToggle', 'checkbox' ])).checkbox,
    ...await getModule([ 'itemToggle', 'checkbox' ]),
    ...await getModule([ 'checkboxElement' ])
  };
});

module.exports = class CheckboxItem extends React.PureComponent {
  constructor (props) {
    super(props);

    this.state = {
      active: props.defaultState
    };
  }

  onClick () {
    const active = !this.state.active;

    this.props.onToggle(active);
    this.setState({ active });
  }

  render () {
    const checkbox = (
      <div
        className={[ classes.item, classes.itemToggle, classes.clickable ].filter(Boolean).join(' ')}
        onClick={() => this.onClick()}
        role='button'
      >
        <div className={classes.label}>
          {this.props.name}
        </div>

        <div className={[ classes.checkbox, classes.itemCheckbox ].filter(Boolean).join(' ')} role='button'>
          <div className={classes.checkboxInner}>
            <input className={classes.checkboxElement} type='checkbox' checked={this.state.active} />
            <span />
          </div>
        </div>
      </div>
    );

    if (this.props.seperate) {
      return React.createElement(ItemGroup, {
        children: [ checkbox ]
      });
    }

    return checkbox;
  }
};
