// powercord/components/ContextMenu/ContextMenu.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2020  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { React, getModule } = require('powercord/webpack');

const ButtonItem = require('./Button');
const CheckboxItem = require('./Checkbox');
const SliderItem = require('./Slider');
const SubmenuItem = require('./Submenu');
const ItemGroup = require('./ItemGroup');

let classes = null;
setImmediate(async () => {
  classes = {
    itemScroller: (await getModule([ 'itemToggle', 'checkbox' ])).scroller,
    ...await getModule([ 'itemToggle', 'checkbox' ]),
    ...await getModule([ 'scrollbar', 'scrollerWrap' ])
  };
});

// @todo: rewrite and extend from base component
module.exports = class ContextMenu extends React.Component {
  render () {
    const children = this.props.itemGroups.map(items => (
      React.createElement(ItemGroup, {
        children: items.map(item => {
          switch (item.type) {
            case 'slider':
              return <SliderItem {...item} />;

            case 'button':
              return <ButtonItem {...item} />;

            case 'submenu':
              return <SubmenuItem {...item} />;

            case 'checkbox':
              return <CheckboxItem {...item} />;

            default:
              return null;
          }
        })
      })
    ));

    return this.renderContextMenu(children);
  }

  renderContextMenu (children) {
    const styles = {
      bottom: `${(window.innerHeight + 24) - this.props.pageY}px`,
      minWidth: this.props.width
    };

    const contextMenuWithScroller = (
      <div className={[ classes.scrollerWrap, classes.scrollerThemed, classes.themeGhostHairline ].filter(Boolean).join(' ')}>
        <div className={[ classes.scroller, classes.systemPad, classes.contextMenu, classes.itemScroller ].filter(Boolean).join(' ')} style={styles}>
          {children}
        </div>
      </div>
    );

    if (this.props.scroller === false) {
      return (
        <div className={classes.contextMenu} style={styles}>
          {children}
        </div>
      );
    }

    return contextMenuWithScroller;
  }
};
