// powercord/components/ContextMenu/Slider.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2019  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { AsyncComponent } = require('powercord/components');
const { React, getModule, getModuleByDisplayName } = require('powercord/webpack');

const Slider = AsyncComponent.from(getModuleByDisplayName('Slider'));

let oldState = null;

module.exports = class SliderItem extends React.PureComponent {
  constructor (props) {
    super();

    this.state = oldState || {
      itemClasses: props.itemClasses || '',
      percentage: props.initialValue
    };
  }

  async componentDidMount () {
    const itemClasses = (await getModule([ 'itemToggle', 'checkbox' ]));

    if (!this.state.itemClasses) {
      this.setState({ itemClasses });
    }
  }

  componentWillUnmount () {
    oldState = this.state;
  }

  render () {
    const { itemClasses } = this.state;
    const slider = (
      <div className={[ itemClasses.item, itemClasses.itemSlider ].join(' ')}>
        <div className={itemClasses.label}>{this.props.name}</div>

        <Slider
          mini={this.props.mini !== void 0 ? this.props.mini : true}
          handleSize={this.props.handleSize}
          className={`${itemClasses.slider} ${this.props.className || ''}`}
          fillStyles={this.props.color ? { backgroundColor: this.props.color } : {}}
          initialValue={this.props.initialValue}
          minValue={this.props.minValue}
          maxValue={this.props.maxValue}
          disabled={this.props.disabled}
          equidistant={this.props.equidistant}
          stickToMarkers={this.props.stickToMarkers}
          markers={this.props.markers}
          onMarkerRender={this.props.onMarkerRender}
          onValueChange={this.props.onValueChange}
          onValueRender={this.props.onValueRender}
        />
      </div>
    );

    if (this.props.seperate) {
      return (
        <div className={`${itemClasses.itemGroup} seperated`}>
          {slider}
        </div>
      );
    }

    return slider;
  }
};
