// powercord/components/ContextMenu/Slider.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2020  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { AsyncComponent } = require('powercord/components');
const { React, getModuleByDisplayName } = require('powercord/webpack');

const Base = require('./Base');
const Slider = AsyncComponent.from(getModuleByDisplayName('Slider'));

module.exports = class SliderItem extends Base {
  render () {
    const { props } = this;
    this.component = React.createElement('div', {
      className: this.classes.itemSlider
    }, React.createElement('span', {
      className: this.classes.label,
      style: props.markers && { marginBottom: '16px' }
    }, props.name), this.renderSlider());

    return super.render();
  }

  renderSlider () {
    return React.createElement(Slider, {
      mini: true,
      className: this.classes.slider,
      fillStyles: this.props.color && { backgroundColor: this.props.color },
      equidistant: typeof this.props.markers !== 'undefined',
      stickToMarkers: typeof this.props.markers !== 'undefined',
      ...this.props
    });
  }
};
