// ac/components/ContextMenu/Submenu.jsx
const { React } = require('ac/webpack');
const Button = require('./Button.jsx');
const Slider = require('./Slider.jsx');

let ContextMenu;

class SubmenuItem extends React.Component {
  constructor () {
    super();

    ContextMenu = require('./ContextMenu.jsx');

    this.state = {
      hovered: false,
      items: [],
      coords: {
        x: 0,
        y: 0
      }
    };
  }

  onEnter (event) {
    const { x, y, height } = event.target.getBoundingClientRect();

    this.setState({
      hovered: true,
      coords: {
        x,
        y: y + height + 2
      }
    });
  }

  onLeave (ev) {
    this.setState({
      hovered: false
    });
  }

  async componentDidMount () {
    const items = await this.props.item.getItems();
    return this.setState({ items });
  }

  render () {
    return (
      <div
        class='item-1Yvehc itemSubMenu-1vN_Yn itemSubMenuHasScroller-1J2hmF'
        onMouseEnter={this.onEnter.bind(this)}
        onMouseLeave={this.onLeave.bind(this)}
      >
        {this.props.item.name}
        {
          this.state.hovered && this.state.items[0]
            ? this.getSubContextMenu()
            : <div />
        }
      </div>
    );
  }

  getSubContextMenu () {
    return (
      <ContextMenu
        pageX={this.state.coords.x}
        pageY={this.state.coords.y}
        width={this.props.item.width || ''}
        itemGroups={[this.state.items]}
      />
    );
  }
}

module.exports = SubmenuItem;
