// powercord/Plugin.js
const Events = require('events');
const { sleep } = require('powercord/util');

const SettingsManager = require('../../Powercord/settingsManager');

module.exports = class Plugin extends Events {
  constructor (options = {}) {
    super();

    this.options = {
      dependencies: [],
      appMode: 'app',
      ...options
    };

    this.ready = false;
    this.settingsManager = new SettingsManager(this.pluginId);
  }

  async _start () {
    try {
      while (!this.options.dependencies.every(pluginName => (
        powercord.pluginManager.get(pluginName).ready
      ))) {
        await sleep(1);
      }

      await this.start();
      this.ready = true;
      this.log('Plugin started');
    } catch (e) {
      this.error('An error occurred during initialization!', e);
    }
  }

  log (...data) {
    console.log(`%c[Powercord:${this.constructor.name}]`, 'color: #257dd4', ...data);
  }

  warn (...data) {
    console.warn(`%c[Powercord:${this.constructor.name}]`, 'color: #257dd4', ...data);
  }

  error (...data) {
    console.error(`%c[Powercord:${this.constructor.name}]`, 'color: #257dd4', ...data);
  }
};
