// powercord/compilers/index.js
/**
 * Powercord, a lightweight @discord client mod focused on simplicity and performance
 * Copyright (C) 2018-2020  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const NoopCompiler = require('./noop');
const JsxCompiler = require('./jsx');
const ScssCompiler = require('./scss');
const LessCompiler = require('./less');
const StylusCompiler = require('./stylus');

module.exports = {
  noop: NoopCompiler,
  jsx: JsxCompiler,
  scss: ScssCompiler,
  less: LessCompiler,
  stylus: StylusCompiler,
  resolveCompiler: (file) => {
    switch (file.split('.').pop()) {
      case 'jsx':
        return new JsxCompiler(file);
      case 'scss':
        return new ScssCompiler(file);
      case 'less':
        return new LessCompiler(file);
      case 'styl':
      case 'stylus':
        return new StylusCompiler(file);
      default:
        return new NoopCompiler(file);
    }
  }
};
