// powercord/compilers/scss.js
/**
 * Powercord, a lightweight @discord client mod focused on simplicity and performance
 * Copyright (C) 2018-2020  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const sass = require('sass');
const Compiler = require('./compiler');
const { existsSync } = require('fs');
const { readFile } = require('fs').promises;
const { join, dirname } = require('path');

// @todo: Remove console.time stuff

/**
 * SCSS compiler
 * @extends {Compiler}
 */
class SCSS extends Compiler {
  async listFiles () {
    console.time('scss imports resolve');
    this.__resolvedImports = {};
    const resolvedFiles = await this._resolveDeps(this.file);
    const res = [
      this.file,
      ...resolvedFiles
    ];
    console.timeEnd('scss imports resolve');
    return res;
  }

  async compile () {
    console.time('scss compile');
    const res = await super.compile();
    console.timeEnd('scss compile');
    return res;
  }

  async computeCacheKey () {
    console.time('scss cache key computation');
    const res = await super.computeCacheKey();
    console.timeEnd('scss cache key computation');
    return res;
  }

  async _compile () {
    console.time('scss render');
    const res = await new Promise((resolve, reject) => {
      readFile(this.file, 'utf8').then(rawScss => {
        sass.render({
          data: rawScss,
          importer: (url, _, callback) => {
            const imported = this.__resolvedImports[url];
            readFile(imported, 'utf8').then(contents => callback({ contents }));
          }
        }, (err, compiled) => {
          if (err) {
            return reject(err);
          }

          resolve(compiled.css.toString());
        });
      });
    });
    console.timeEnd('scss render');
    return res;
  }

  /**
   * Resolve dependencies imported in SCSS files.
   * @param {String} file File to crawl
   * @returns {Promise<String[]>}
   */
  async _resolveDeps (file, resolvedFiles = []) {
    const scss = await readFile(file, 'utf8');
    const basePath = dirname(file);
    const foundFiles = [];
    // Maybe in the future show a deprecation notice for the use of import (?)
    for (const match of scss.matchAll(/@(?:import|use) ['"]([^'"]+)/ig)) {
      let filePath = join(basePath, match[1]).replace(/\\/g, '/');
      if (!filePath.endsWith('.scss')) {
        filePath += '.scss';
      }
      if (!existsSync(filePath)) {
        const f = filePath.split('/');
        f[f.length - 1] = `_${f[f.length - 1]}`;
        filePath = f.join('/');
        if (!existsSync(filePath)) {
          throw new Error(`Cannot find imported file "${match[1]}".`);
        }
      }
      if (!foundFiles.includes(filePath)) {
        this.__resolvedImports[match[1]] = filePath;
        foundFiles.push(filePath);
      }
    }

    for (const foundFile of foundFiles) {
      if (!resolvedFiles.includes(foundFile)) {
        resolvedFiles.push(foundFile);
        await this._resolveDeps(foundFile, resolvedFiles);
      }
    }

    return resolvedFiles;
  }

  get _metadata () {
    return `${sass.info}; Powercord import resolver v1`;
  }
}

module.exports = SCSS;
