// powercord/components/settings/ButtonItem.jsx
const { getModule, getModuleByDisplayName, React } = require('powercord/webpack');
const AsyncComponent = require('../AsyncComponent');

const DFormItem = AsyncComponent.from(getModuleByDisplayName('FormItem'));
const FormDivider = AsyncComponent.from(getModuleByDisplayName('FormDivider'));
const FormText = AsyncComponent.from(getModuleByDisplayName('FormText'));
const Button = AsyncComponent.from(getModuleByDisplayName('DropdownButton'));

let classes = {
  initialized: false,
  flexClassName: '',
  classMargins: {},
  classTitle: '',
  classDescription: ''
};

module.exports = class ButtonItem extends React.Component {
  constructor () {
    super();

    this.state = { classes };
  }

  async componentDidMount () {
    if (classes.initialized) {
      return;
    }

    const Flex = await getModuleByDisplayName('Flex');
    classes = {
      initialized: true,

      flexClassName: `${Flex.Direction.VERTICAL} ${Flex.Justify.START} ${Flex.Align.STRETCH} ${Flex.Wrap.NO_WRAP}`,
      classMargins: await getModule([ 'marginTop20' ]),
      classTitle: (await getModule([ 'titleDefault' ])).titleDefault,
      classDescription: (await getModule([ 'formText', 'description' ])).description
    };

    this.setState({ classes });
  }

  render () {
    return <DFormItem
      className={`powercord-settings-item powercord-buttonitem ${this.state.classes.flexClassName} ${this.state.classes.classMargins.marginBottom20}`}>
      <div className='powercord-settings-item-title'>
        <div>
          <div className={this.state.classes.classTitle}>
            {this.props.children}
          </div>
          <FormText className={this.state.classes.classDescription}>
            {this.props.note}
          </FormText>
        </div>
        <Button onClick={() => this.props.onClick()}>{this.props.button}</Button>
      </div>
      <FormDivider className={this.state.classes.classMargins.marginTop20}/>
    </DFormItem>;
  }
};
