// powercord/components/settings/FormItem.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2019  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { getModule, getModuleByDisplayName, React } = require('powercord/webpack');
const AsyncComponent = require('../AsyncComponent');

const DFormItem = AsyncComponent.from(getModuleByDisplayName('FormItem'));
const FormDivider = AsyncComponent.from(getModuleByDisplayName('FormDivider'));
const FormText = AsyncComponent.from(getModuleByDisplayName('FormText'));

let classes = {
  initialized: false,
  flexClassName: '',
  classModule: {},
  formModule: {}
};

module.exports = class FormItem extends React.Component {
  constructor () {
    super();

    this.state = { classes };
  }

  async componentDidMount () {
    if (classes.initialized) {
      return;
    }

    const Flex = await getModuleByDisplayName('Flex');
    classes = {
      initialized: true,

      flexClassName: `${Flex.Direction.VERTICAL} ${Flex.Justify.START} ${Flex.Align.STRETCH} ${Flex.Wrap.NO_WRAP}`,
      classModule: await getModule([ 'marginTop20' ]),
      formModule: (await getModule([ 'formText', 'description' ]))
    };

    this.setState({ classes });
  }

  render () {
    return (
      <DFormItem
        title={this.props.title}
        required={this.props.required}
        className={`${classes.flexClassName} ${classes.classModule.marginBottom20}`}
      >
        {this.props.children}

        <FormText className={classes.formModule.description}>
          {this.props.note}
        </FormText>

        <FormDivider className={classes.classModule.marginTop20} />
      </DFormItem>
    );
  }
};
