// powercord/components/settings/KeybindRecorder.jsx
/**
 * Powercord, a lightweight @discord client mod focused on simplicity and performance
 * Copyright (C) 2018-2020  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { getModuleByDisplayName, React } = require('powercord/webpack');
const AsyncComponent = require('../AsyncComponent');
const FormItem = require('./FormItem.jsx');

const Button = AsyncComponent.from(getModuleByDisplayName('DropdownButton'));

const renameKeys = {
  ' ': 'Space',
  Control: 'Ctrl',
  AltGraph: 'AltGr',
  ArrowUp: 'Up',
  ArrowDown: 'Down',
  ArrowLeft: 'Left',
  ArrowRight: 'Right',
  AudioVolumeUp: 'VolumeUp',
  AudioVolumeDown: 'VolumeDown',
  AudioVolumeMute: 'VolumeMute'
};

module.exports = class TextInput extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      id: `pc-kbr-${Date.now()}-${Math.round(Math.random() * 1000)}`,
      recording: false,
      value: this.props.value,
      listeners: null
    };

    this.forceUpdate = false;
  }

  componentDidUpdate (prevProps) {
    if (this.props.value !== prevProps.value || this.forceUpdate) {
      this.forceUpdate = false;
      this.setState({ value: this.props.value });
    }
  }

  render () {
    const className = this.state.recording ? 'powercord-keybind recording pc-container' : 'powercord-keybind pc-container pc-hasValue';
    return (
      <FormItem title={this.props.children} note={this.props.note} required={this.props.required}>
        <div className='powercord-keybind-container'>
          <div className={className} onClick={(e) => {
            e.preventDefault();
            e.stopPropagation();
            this.setRecording(!this.state.recording);
          }}>
            <div className='powercord-keybind-inner pc-layout pc-flex'>
              <input
                className='powercord-keybind-input pc-input pc-base'
                value={this.state.value}
                id={this.state.id}
              />
              <div className='powercord-keybind-btn pc-flex'>
                <button type='button' className='powercord-keybind-button pc-button pc-grow'>
                  <div className='powercord-keybind-button-inner pc-contents'>
                    <span className='text pc-text'>{this.state.recording ? 'Stop Recording' : 'Edit Keybind'}</span>
                    <span className='icon pc-icon' />
                  </div>
                </button>
              </div>
            </div>
          </div>
          <Button onClick={() => this.props.onReset()}>Reset keybind</Button>
        </div>
      </FormItem>
    );
  }

  componentWillUnmount () {
    this.setRecording(false);
  }

  setRecording (recording) {
    const element = document.getElementById(this.state.id);
    if (recording && !this.state.recording) {
      const inputListener = (e) => {
        e.preventDefault();
        e.stopPropagation();

        if (e.repeat) {
          return;
        }

        const rawKey = e.key[0].toUpperCase() + e.key.slice(1);
        const key = renameKeys[rawKey] || rawKey;

        if ((this.state.value.match(/\+\+?/g) || []).length !== 3) {
          this.setState({ value: this.state.value === '' ? key : `${this.state.value}+${key}` });
        }

        if ((`${this.state.value}+${key}`.match(/\+\+?/g) || []).length === 3) {
          element.blur();
          this.setRecording(false);
        }
      };

      const clickListener = () => {
        this.setRecording(false);
      };

      element.focus();
      element.addEventListener('keydown', inputListener);
      window.addEventListener('click', clickListener);

      if (this.props.onRecord) {
        this.props.onRecord();
      }

      this.setState({
        recording: true,
        value: '',
        listeners: {
          input: inputListener,
          window: clickListener
        }
      });
    } else if (!recording && this.state.recording) {
      element.removeEventListener('keydown', this.state.listeners.input);
      window.removeEventListener('click', this.state.listeners.window);

      this.forceUpdate = true;
      if (this.state.value === '') {
        this.props.onReset();
      } else {
        this.props.onChange(this.state.value);
      }

      this.setState({
        recording: false,
        listeners: null
      });
    }
  }
};
