// powercord/components/index.js
const { getModuleByDisplayName } = require('powercord/webpack');
const AsyncComponent = require('./AsyncComponent');

require('fs')
  .readdirSync(__dirname)
  .filter(file => file !== 'index.js')
  .forEach(filename => {
    const moduleName = filename.split('.')[0];
    exports[moduleName] = require(`${__dirname}/${filename}`);
  });

// Add some util components
Object.assign(exports, {
  Icon: AsyncComponent.from(getModuleByDisplayName('Icon')),
  Button: AsyncComponent.from(getModuleByDisplayName('DropdownButton')),

  /**
   * @todo pretty self explanatory, replace with native component
   * when discord eventually axes it
   */
  Tooltip: AsyncComponent.from(getModuleByDisplayName('TooltipDeprecated')),
  Switch: AsyncComponent.from(getModuleByDisplayName('Switch')),
  Spinner: AsyncComponent.from(getModuleByDisplayName('Spinner')),
  Divider: AsyncComponent.from(getModuleByDisplayName('FormDivider'))
});

// Re-export module properties
(async () => {
  const Button = await getModuleByDisplayName('DropdownButton', true, true);
  const Icon = await getModuleByDisplayName('Icon', true, true);
  [ 'DropdownSizes', 'Sizes', 'Colors', 'Looks' ].forEach(prop => exports.Button[prop] = Button[prop]);
  [ 'Names' ].forEach(prop => exports.Icon[prop] = Icon[prop]);
})();
