// powercord/components/Toast.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2019  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { React, getModule, getModuleByDisplayName } = require('powercord/webpack');
const AsyncComponent = require('./AsyncComponent');

const Clickable = AsyncComponent.from(getModuleByDisplayName('Clickable'));
const Tooltip = AsyncComponent.from(getModuleByDisplayName('TooltipDeprecated'));

Object.assign(exports, {
  Button: AsyncComponent.from(getModule(m => m.name === 'm'))
});

(async () => {
  const Button = await getModule(m => m.name === 'm', true, true);
  [ 'Colors', 'Hovers', 'Looks', 'Sizes' ].forEach(prop => exports.Button[prop] = Button[prop]);
})();

/* eslint-disable multiline-ternary */
const Toast = class Toast extends React.PureComponent {
  constructor (props) {
    super(props);

    this.state = {
      submitting: [],
      leaving: false
    };

    this.styles = {
      regular: 'r',
      light: 'l',
      duotone: 'd',
      brands: 'b'
    };
  }

  render () {
    const { Button } = exports;
    let dismissTimeout;

    if (this.props.timeout && !isNaN(this.props.timeout)) {
      dismissTimeout = setTimeout(() => powercord.api.toasts.closeToast(this.props.id), this.props.timeout);
    }

    return (
      <div
        id={this.props.id}
        className={[ 'powercord-toast', this.isLeaving() ? 'leaving' : '' ].filter(Boolean).join(' ')}
        data-toast-type={this.props.type || 'info'}
        style={this.props.style}
      >
        {this.props.header && (
          <div className='header'>
            {this.props.icon !== false && (
              <Tooltip
                text={`${this.props.type
                  ? this.props.type.replace(/\w\S*/g, (text) => text.charAt(0).toUpperCase() + text.substr(1).toLowerCase())
                  : 'Info'}: ${this.props.header}`}
                position='left'
              >
                <div className='icon' style={this.props.iconColor ? { color: this.props.iconColor } : {}}>
                  {this.props.image ? this.getHeaderImage() : this.props.icon
                    ? this.getHeaderIcon()
                    : <div className={`far fa-${this.props.type && this.objectHasKey(this.props.type, Toast.Types)
                      ? Toast.Types[(this.props.type).toUpperCase()]
                      : Toast.Types.INFO} fa-fw`} />}
                </div>
              </Tooltip>
            )}

            {this.props.header}

            <Clickable
              className='dismiss'
              onClick={() => {
                clearTimeout(dismissTimeout);
                powercord.api.toasts.closeToast(this.props.id);
              }}
            >
              <div className='fal fa-times fa-fw' />
            </Clickable>
          </div>
        )}

        {(this.props.content || (this.props.buttons && this.props.buttons.filter(Boolean).length > 0)) && (
          <div className='body'>
            {this.props.content && (
              <div className='content'>
                {this.props.content}
              </div>
            )}

            <div className='buttons'>
              {Array.isArray(this.props.buttons) && (
                this.props.buttons.filter(Boolean).map(button => (
                  <Button
                    className='button'
                    submitting={this.state.submitting.includes(button.text)}
                    size={button.size && this.objectHasKey(button.size, Button.Sizes)
                      ? Button.Sizes[button.size.toUpperCase()]
                      : Button.Sizes.SMALL}
                    color={button.danger ? Button.Colors.RED : button.color
                      ? Button.Colors[typeof button.color === 'string' ? button.color.toUpperCase() : '']
                      : Button.Colors.TRANSPARENT}
                    look={button.danger ? Button.Looks.OUTLINED : button.type && this.objectHasKey(button.type, Button.Looks)
                      ? Button.Looks[(button.type).toUpperCase()]
                      : Button.Looks.FILLED}
                    hover={button.hoverColor
                      ? Button.Hovers[typeof button.hoverColor === 'string' ? (button.hoverColor).toUpperCase() : '']
                      : Button.Hovers.TRANSPARENT}
                    style={button.color ? this.getButtonStyle(button) : null}
                    onClick={() => {
                      if (button.onClick) {
                        button.onClick();
                      }

                      clearTimeout(dismissTimeout);
                      this.state.submitting.push(button.text);

                      return powercord.api.toasts.closeToast(this.props.id);
                    }}
                  >
                    {button.text}
                  </Button>
                ))
              )}
            </div>
          </div>
        )}
      </div>
    );
  }

  getHeaderImage () {
    return <img alt='' className={[ this.props.imageClassName || null ].filter(Boolean).join(' ')}
      src={this.props.image} />;
  }

  getHeaderIcon () {
    const styleRegex = new RegExp(/[a-z]+(?!.*-)/);
    const style = Object.keys(this.styles).find(style => style === this.props.icon.split(' ')[0].match(styleRegex)[0]);
    const icon = `fa-${this.props.icon.replace(`-${style}`, '')} fa-fw`;

    let prefix = 'fas';

    if (this.styles[style]) {
      prefix = `fa${this.styles[style]}`;
    }

    return <div className={`${prefix} ${icon}`} />;
  }

  getButtonStyle (button) {
    let { style } = button;
    const { isValidHex } = getModule([ 'hex2rgb' ], false);

    if (isValidHex(button.color)) {
      style = { backgroundColor: button.color };
    } else {
      let foreground, background;

      if (Array.isArray(button.color)) {
        [ foreground, background ] = button.color;
      } else if (typeof button.color === 'object') {
        ({ foreground, background } = button.color);

        if (button.color.primary || button.color.secondary) {
          ({ primary: foreground, secondary: background } = button.color);
        }
      }

      return { color: foreground && isValidHex(foreground) ? foreground : null,
        borderColor: background && isValidHex(background) ? background : null };
    }

    return style;
  }

  objectHasKey (value, object) {
    if (object.hasOwnProperty(value.toUpperCase())) {
      return true;
    }

    return false;
  }

  isLeaving () {
    const toast = powercord.api.toasts.toasts.find(toast => toast.id === this.props.id);
    if (toast && toast.leaving) {
      return true;
    }

    return false;
  }
};

Toast.Types = {
  INFO: 'info-circle',
  WARN: 'exclamation-circle',
  WARNING: 'exclamation-circle',
  ERROR: 'times-circle',
  DANGER: 'times-circle',
  SUCCESS: 'check-circle'
};

module.exports = Toast;
