// powercord/components/Toast.jsx
const { React } = require('powercord/webpack');

module.exports = class Toast extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      header: props.header,
      buttons: props.buttons,
      content: props.content
    };
  }

  onBtnClick (onClick) {
    return onClick(state => Promise.resolve(this.setState(state)));
  }

  render () {
    const { props, state } = this;

    return (
      <div
        className={`powercord-toast${state.leaving ? ' leaving' : ''}`}
        style={props.style}
      >
        {state.header && (
          <div className='header'>
            {state.header}
          </div>
        )}

        <div className={state.fade ? `body fade${state.fade}` : 'body'}>
          {state.content && (
            <div className='content'>
              {state.content}
            </div>
          )}

          <div className='btns'>
            {state.buttons.map(btn => (
              <div className='btn' onClick={this.onBtnClick.bind(this, btn.onClick)}>
                {btn.text}
              </div>
            ))}
          </div>
        </div>
      </div>
    );
  }
}
