// powercord/components/Toast.jsx
/**
 * Powercord, a lightweight @discordapp client mod focused on simplicity and performance
 * Copyright (C) 2018-2019  aetheryx & Bowser65
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const { React } = require('powercord/webpack');
const { sleep } = require('powercord/util');

module.exports = class Toast extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      header: props.header,
      buttons: props.buttons,
      content: props.content
    };
  }

  onBtnClick (onClick) {
    return onClick(state => new Promise(async (resolve) => {
      this.setState(state);

      if (state.fade) {
        await sleep(200);
      }
      if ('leaving' in state) {
        await sleep(500);
      }

      resolve();
    }));
  }

  render () {
    const { props, state } = this;

    return (
      <div
        className={`powercord-toast${state.leaving ? ' leaving' : ''}`}
        style={props.style}
      >
        {state.header && (
          <div className='header'>
            {state.header}
          </div>
        )}

        <div className={state.fade ? `body fade${state.fade}` : 'body'}>
          {state.content && (
            <div className='content'>
              {state.content}
            </div>
          )}

          <div className='buttons'>
            {state.buttons.map(button => (
              <div className='button' onClick={this.onBtnClick.bind(this, button.onClick)}>
                {button.text}
              </div>
            ))}
          </div>
        </div>
      </div>
    );
  }
};
